docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc

```diff
@@ -1110,6 +1110,111 @@ The default tenant configuration is automatically disabled when `quarkus.oidc.au
 
 Be aware that tenant-specific configurations can also be disabled, for example: `quarkus.oidc.tenant-a.tenant-enabled=false`.
 
+[[programmatic-startup]]
+== Programmatic OIDC start-up for multitenant application
+
+Static OIDC tenants can be created programmatically like in the example below:
+
+[source,java]
+----
+package io.quarkus.it.oidc;
+
+import io.quarkus.oidc.Oidc;
+import io.quarkus.oidc.OidcTenantConfig;
+import jakarta.enterprise.event.Observes;
+
+public class OidcStartup {
+
+    void observe(@Observes Oidc oidc) { <1>
```

Comment:
Should it really be an event as well ? IMHO having it injected is better, and used in on start up events. It may be a bit confusing otherwise.

By the way, how does using `Oidc` work in `main()` ?

Reply:
The question is probably the most important part - it is really important that we don't pledge ourselves to start Oidc after all the startup events has been fired, because we don't know what else are users doing during the startup that may require Oidc. If we agree that users can create static tenants with the startup event, we also agree to start Oidc after all the startup events has been processed (sync or async) but we need to do it before HTTP server has started. After all the startup events has been fired, only thing that I know that happens is starting the HTTP server (sure, I could be wrong) and we will need to get somewhere in between. I think it is better to own our own timing.

I'll throw few arguments because this is topic I have thought about a lot. Please let me know if they don't make sense:

1. We fire this when bean container is completely ready.
2. We fire this when runtime config is completely ready.
3. We need `TenantConfigBean` ready before HTTP router is finalized (no the same as HTTP server started) before than, even with `Router` is fired and `io.quarkus.oidc.runtime.BackChannelLogoutHandler` needs to use it.  BTW why shouldn't we send `Oidc` as event when other events like `Router` are acceptable?

Reply:
> By the way, how does using Oidc work in main() ?

Do you mean this https://quarkus.io/guides/lifecycle#the-main-method? Docs says `It is not recommenced to do any business logic in this main method, as Quarkus has not been set up yet` so no, it won't and shouldn't.

Reply:
> IMHO having it injected is better

We need to differ between static and dynamic tenants. The static ones are final and we must have them before Oidc starts. You can inject `Oidc` and create dynamic tenants.

Reply:
Not sure if it is clear why I think this is so important. And I can be wrong...  Just for case - some features are only available for the static tenants and they are setup once (like issue/path-based resolvers, backchannel logout etc.) and we need to have static tenants ready when we do that. If we allow users to create static tenants anytime, we will have to refactor a lot of things and I think it will give only little fruit. We need static tenants ready before users use OIDC and if we own logic when we fire the event with the `OIDC`, then we have flexibility. We can always postpone it when necessary as long as the HTTP server is not running. But once we say: _you can create static tenants in the startup events_, we will never be able to move if before that without potentially breaking changes.

Reply:
Hi Michal, @michalvavrik, so `Oidc` is the startup kind of event, right ? (And sure, if it also works for `Router` then it only makes sense to allow it for `Oidc`).

I'd only ask for us not to go into the dynamic OIDC tenants space at this stage.

As far as I understand, the main reason some users prefer the programmatic approach is because they don't want properties. And we already let users create dynamic tenants with `TenantConigResolver`.

So, it appears to me the main use case, right now at least, is to create, possibly a single (static) tenant programmatically at startup.

We can definitely evolve and grow it further, but I'd start with the simplest possible option

Thanks

Reply:
Cool @sberyozkin ; I also think the main use case are static tenants. I can propose allowing registering dynamic tenants  later together with the `@Tenant` annotation. Static tenants were my main motivation. I'll put to it to my list and propose something in regards of dynamic tenants in the second half of January. It should make this PR little bit easier to review.

Reply:
> Hi Michal, @michalvavrik, so Oidc is the startup kind of event, right ?

Yes, I suspect it will happen right before the startup event, but I don't have working knowledge of build steps ordering, there is some asynchronous mechanism I have never checked.